python/src/iceberg/expressions/base.py

```diff
@@ -346,3 +346,116 @@ def bind(self, schema: Schema, case_sensitive: bool) -> BoundReference:
             raise ValueError(f"Cannot find field '{self.name}' in schema: {schema}")
 
         return BoundReference(field=field, accessor=schema.accessor_for_field(field.field_id))
+
+
+class BooleanExpressionVisitor(ABC):
+    @abstractmethod
+    def visit_always_true(self) -> BooleanExpression:
+        """Visit method for an AlwaysTrue boolean expression
+
+        Note: This visit method has no arguments since AlwaysTrue instances have no context.
+
+        Returns:
+            BooleanExpression: The resulting expression after any modifications performed by the visitor
+        """
+
+    @abstractmethod
+    def visit_always_false(self) -> BooleanExpression:
+        """Visit method for an AlwaysFalse boolean expression
+
+        Note: This visit method has no arguments since AlwaysFalse instances have no context.
+
+        Returns:
+            BooleanExpression: The resulting expression after any modifications performed by the visitor
+        """
+
+    @abstractmethod
+    def visit_not(self, result: BooleanExpression) -> BooleanExpression:
```

Comment:
The result should have type `T` because this is passed the result of visiting the child.

Reply:
Updated this to `visit_not(self, child_result: T) -> T` 👍